boot/fdehook/fdehook.go

```diff
@@ -0,0 +1,131 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package fdehook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+func fdeHook(kernelDir string) string {
+	return filepath.Join(kernelDir, "meta/hooks/fde")
+}
+
+// Enabled returns whether the external FDE helper should be called
+func Enabled(kernelDir string) bool {
+	return osutil.FileExists(fdeHook(kernelDir))
+}
+
+// fdeHookCmd returns a *exec.Cmd that runs the fdehook code with
+// (some) sandboxing around it. The sandbox will ensure that it's hard
+// for the hook to abuse that they are called in e.g. initrd.  But it
+// does not aim for perfect protection - the fdehooks are part of the
+// kernel snap/initrd so if someone wants to do mischief there are
+// easier ways by hacking initrd directly.
+func fdeHookCmd(kernelDir string, args ...string) *exec.Cmd {
+	cmd := exec.Command(
+		"systemd-run",
```

Comment:
I was thinking if this should be done over the DBus API but I think this is sufficient and the extra preparation by `systemd-run` is non-trivial enough to keep this as a subprocess call. I would reverse this only if we need to monitor the precise exit status of the original process.

Reply:
Added a comment there.